### DISCUSSION TITLE
Questions related to Cilium Clustermesh/Kvstoremesh

### DISCUSSION THREAD

**Discussion by krunaljain** [cilium/cilium#35746] [Q&A] [answered]
Hi. Had a few questions related to clustermesh and kvstoremesh. If anyone has context, can you please help with the answers?

- How are the credentials to connect to the remote clusters populated under the clustermesh config directory?
- In kvstoremesh, the clustermesh apiserver connects to remote etcd instances. Does it use the same credentials to do so?
- Does clustermesh enablement trigger a cilium agent restart for the cluster? Does any other clustermesh related operation trigger an agent restart?
- For connecting two clusters, both of them need to have the same datapath mode. What is the main reason enforcing this limitation?

**Comment by giorio94** [comment, marked as answer]
> How are the credentials to connect to the remote clusters populated under the clustermesh config directory?

Clustermesh-related configurations and credentials are mounted from the `cilium-clustermesh`, `clustermesh-apiserver-local-cert` and `clustermesh-apiserver-remote-cert` secrets. Assuming that the same root CA is configured in all clusters, the `cilium-clustermesh` secret essentially contains the IP/address+port of the etcd instance of each target cluster, while each cluster uses the locally generated credentials contained in the other secrets (i.e., private keys and certificates don't need to be copied from one cluster to another).

> In kvstoremesh, the clustermesh apiserver connects to remote etcd instances. Does it use the same credentials to do so?

Yes, it uses the same credentials that agents would use if kvstoremesh were disabled. Starting from v1.17 \[1], Cilium agents use different credentials (associated with different permissions) to connect to the local kvstoremesh instance though, while historically they used the same here as well.

> Does clustermesh enablement trigger a cilium agent restart for the cluster? Does any other clustermesh related operation trigger an agent restart?

Agent (and clustermesh-apiserver, if kvstoremesh is enabled) restart is currently triggered when connecting to/disconnecting from a cluster targeted by IP address (and not FQDN address). This is because that triggers the addition/removal of an `hostAlias` entry in the DaemonSet (and Deployment) \[2]. This may change in the future though to avoid this need.

> For connecting two clusters, both of them need to have the same datapath mode. What is the main reason enforcing this limitation?

Essentially the datapath does not make any particular distinction whether a given node belongs to the local or a remote cluster, and it expects that the configuration is coherent (in terms of routing mode, encryption settings, ...).

\[1]: https://github.com/cilium/cilium/pull/33137
\[2]: https://github.com/cilium/cilium/blob/12bbed5aa1b4612aaca6bfac965a27bbb41b3d57/install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml#L781-L787

**Comment by superzhu** [comment]
Borrowing this thread, ask some questions:
1. When I install ClusterMesh with HA mode, for example, the ClusterMesh consists of three instances, do the etcd instances inside ClusterMesh-apiserver pods form an etcd cluster?
2. I read some ClusterMesh related source code, etcd-init part creates some users in its local etcd instance? can I say ClusterMesh only user co-located instance, cannot use externally deploy etcd server?

**Reply by giorio94** [reply]
> When I install ClusterMesh with HA mode, for example, the ClusterMesh consists of three instances, do the etcd instances inside ClusterMesh-apiserver pods form an etcd cluster?

No, each etcd instance is independent, and they don't form a cluster.

> I read some ClusterMesh related source code, etcd-init part creates some users in its local etcd instance? can I say ClusterMesh only user co-located instance, cannot use externally deploy etcd server?

It depends. If Cilium is already configured to use etcd as KVStore (opposed to the default CRD mode), then you don't need the clustermesh-apiserver, and you can simply point the clustermesh configuration of the other clusters to it. In that case, the configuration of authentication/authorization policies is completely on you.

Otherwise, when Cilium runs in CRD mode, you need to use the clustermesh-apiserver to expose the information to the other clusters. You could theoretically configure it to work with an external etcd cluster, rather than the sidecar one, but there's no support in the helm chart for that at the moment, and there may be also some rough edges here and there that would need to be fixed given that it is not tested at all.

**Reply by superzhu** [reply]
@giorio94 thanks for clarification

**Comment by krunaljain** [comment]
Thanks for providing the answers. I had a few more questions related to the clustermesh functionality:

- I observed the helm chart values don't get updated even when I run clustermesh enable and connect commands although the code seems to be updating the chart in the backed?
- Once the agents sync the service, endpoints and identities information from the remote clusters, how is the local cache information applied to the underlying ebpf programs for service discovery and routing?
- Does the cilium agent monitor the clustermesh config directory even if clustemesh is not enabled?

**Reply by giorio94** [reply]
> I observed the helm chart values don't get updated even when I run clustermesh enable and connect commands although the code seems to be updating the chart in the backed?

Weird. Are you using an old CLI version by any chance? Newer versions only use helm under the hood.

> Once the agents sync the service, endpoints and identities information from the remote clusters, how is the local cache information applied to the underlying ebpf programs for service discovery and routing?

The entries from remote clusters get essentially merged with the information from the local cluster, and eventually plumbed into the appropriate BPF maps.

> Does the cilium agent monitor the clustermesh config directory even if clustemesh is not enabled?

Yes, there's no explicit flag enabling clustermesh, besides the config directory being specified, and the cluster ID being not zero.
